autopilot/purepursuitwaypointfollower.cpp

```diff
@@ -523,4 +555,153 @@ QSharedPointer<VehicleState> PurepursuitWaypointFollower::getReferenceVehicleSta
     } else {
         return mVehicleState;
     }
-}
\ No newline at end of file
+}
+
+void PurepursuitWaypointFollower::addSpeedLimitRegion(const QList<PosPoint>& boundary, double maxSpeed_kmph)
+{
+    if (boundary.size() < 3) {
+        qDebug() << "WARNING: Speed limit region requires at least 3 points to define a polygon.";
+        return;
+    }
+
+    double maxSpeed_ms = maxSpeed_kmph / 3.6;  // convert km/h to m/s
+
+    SpeedLimitRegion region;
+    region.boundary = boundary;
+    region.maxSpeed = maxSpeed_ms;
+    mSpeedLimitRegions.append(region);
+
+    qInfo() << "Added speed limit region with" << boundary.size() << "vertices and max speed" << maxSpeed_kmph << "km/h. Keep in mind, any potential inner regions or holes are ignored";
+}
+
+void PurepursuitWaypointFollower::clearSpeedLimitRegions()
+{
+    mSpeedLimitRegions.clear();
+    qDebug() << "Cleared all speed limit regions";
+}
+
+QList<SpeedLimitRegion> PurepursuitWaypointFollower::getSpeedLimitRegions() const
+{
+    return mSpeedLimitRegions;
+}
+
+void PurepursuitWaypointFollower::loadSpeedLimitRegionsFile()
+{
+    QDir documentsDirectory(QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation));   // OS agnostic
+
+    QString folderName = "WayWise Speed Limits";
+    QString folderPath = documentsDirectory.filePath(folderName);
+
+    if (!documentsDirectory.exists(folderPath)) {
+        if (documentsDirectory.mkpath(folderPath)) {
+            qInfo() << "Speed limits folder created";
+        } else {
+            qWarning() << "Failed to create speed limits folder";
+            return;
+        }
+    }
+
+    QString fileName = "speedLimitRegions.json";
+    QString filePath = documentsDirectory.filePath(folderName + "/" + fileName);
+
+    QFile file(filePath);
+    if (!file.open(QIODevice::ReadOnly)) {
+        qInfo() << "Could not open speed limit regions file:" << filePath << Qt::endl
```

Comment:
```suggestion
    QFile file(mSpeedLimitRegionsFilePath);
    if (!file.open(QIODevice::ReadOnly)) {
        qInfo() << "Could not open speed limit regions file:" << mSpeedLimitRegionsFilePath << Qt::endl
```

Instead of a hardcoded filepath, I would recommend adding a mSpeedLimitRegionsFilePath QString to the class along with its setter and getter for better configuration.